sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1201 @@
+"""
+Metadata Routing Utility
+"""
+
+# Author: Adrin Jalali <adrin.jalali@gmail.com>
+# License: BSD 3 clause
+
+import functools
+import inspect
+from copy import deepcopy
+from enum import Enum
+from warnings import warn
+from collections import namedtuple
+from typing import Union, Optional
+from ..externals._sentinels import sentinel  # type: ignore # mypy error!!!
+from ._bunch import Bunch
+
+# This namedtuple is used to store a (mapping, routing) pair. Mapping is a
+# MethodMapping object, and routing is the output of `get_metadata_routing`.
+# MetadataRouter stores a collection of these namedtuples.
+RouteMappingPair = namedtuple("RouteMappingPair", ["mapping", "routing"])
+
+# A namedtuple storing a single method route. A collection of these namedtuples
+# is stored in a MetadataRouter.
+Route = namedtuple("Route", ["method", "used_in"])
```

Comment:
I wonder whether `source` and `dest` (or `src` and `dst`) is clearer, or less so.

Reply:
For me, there is mental overhead in mapping `used_in` and `method` to the correct semantics. (I keep forgetting which is which going through this PR.)

Given that the namedtupled called "Route", I think something like "src" and "dst" is semantically better. Which means:

- `method` -> `dst`
- `used_in` -> `src`

(I hope I got the semantics here correct)

Reply:
I think that's because you're looking at the implementation rather than the code written by a user. If you forget about how it's implemented and look at how it's used, then it makes sense. The user doesn't need to know anything about routing. They just need to know which method of a child object/estimator is used in which method of a parent estimator.

I think when somebody reads this:

``` python
    def get_metadata_routing(self):
        router = (
            MetadataRouter()
            .add(
                transformer=self.transformer,
                method_mapping=MethodMapping()
                .add(method="fit", used_in="fit")
                .add(method="transform", used_in="fit")
                .add(method="transform", used_in="predict"),
            )
            .add(classifier=self.classifier, method_mapping="one-to-one")
        )
        return router.serialize()
```

it's very clear what it does. If you think the keywords need to be more explicit, we could do `child_method` and `parent_method`, but that again requires the user to understand why one is child and another one is a parent.

I'm also happy to rename objects/classes in the implementation to change your mental model when you read it. I think if `Route` was called `MethodPair`, then you wouldn't expect `src` and `dst` maybe?